### PR TITLE
#162170718 Users should be able to receive notifications

### PR DESCRIPTION
**Pull request by akhilome**
#### What does this PR do?

Adds email and in-app notification functionality.

#### Description of Task to be completed?

- write tests and implement functionality to give users
ability to optin and out of email (and or) in-app notifications
- implement email notifications using sendgrid
- save notifications to the database
- implement in app notifications using pusher

#### How should this be manually tested?

- clone the repo, cd into it, and run npm install to install the project's dependencies,
- rename the env.sample file to .env and fill out the required environmental variables,
- start the local server by running npm start, then hit the create an account using the singup endpoint, per the documentation,
- After signing up the first user, sign up another user, and follow the first user using the `POST
/profiles/username/follow` endpoint.
- The first user would receive a notification (email and inapp) about the follow

#### Any background context you want to provide?

N/A

#### What are the relevant pivotal tracker stories? (If applicable)

[#162170718](https://www.pivotaltracker.com/story/show/162170718)

#### Screenshots (If applicable)

image

image

image

#### Questions:

N/A

### CODE REVIEW

File: src/controllers/notificationsController.js [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+   * @param {Object} request - request object
+   * @param {Object} response - reponse object
+   * @returns {Object} - sucess response or error message
+   */
+  static async optInOut(request, response) {
+    const { id } = request.userData.payload;
+    const validQuery = param => (param === 'yes' || param === 'no');
+
+    try {
+      const user = await User.findByPk(id);
```
Comment (okoroemeka) [inline src/controllers/notificationsController.js:24]: Nice Job @akhilome. What does findByPk do?

**Review by Johnsonojo — COMMENTED** [review]
Nice work @akhilome. I Like the way you structured your implementation.

File: src/helpers/notification.js [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+
+/**
+ * @class Notifier
+ */
+class Notifier {
+  /**
+   * @description gets all users from the database who donot want email/app notifications.
+   * @static
+   * @param {string} notifType - type of notification
+   * @returns {object[]} - array of users who donot want to receive app/email notifications.
```
Comment (Johnsonojo) [inline src/helpers/notification.js]: @akhilome, I think there is a typo here

**Review by okoroemeka — APPROVED** [review]

File: src/helpers/notification.js (hunk 2 above)
Comment (akhilome) [reply src/helpers/notification.js]: Thank you, this would be fixed.

File: src/controllers/notificationsController.js (hunk 1 above)
Comment (akhilome) [reply src/controllers/notificationsController.js:24]: Hey @okoroemeka, `findByPk` is used for searching for a single instance by its primary key. It's a replacement for `findById` which is now deprecated.

**Review by Johnsonojo — APPROVED** [review]

**Pull request merged**